### PR TITLE
Replace deprecated Lwt usages

### LINKED ISSUE #13: Move from Lwt_log to the logs library
We currently use the [`Lwt_log`](https://ocsigen.org/lwt/2.7.1/api/Lwt_log) library for logging in this repo. As https://github.com/ocsigen/lwt/issues/468#issuecomment-336575454 said, [`Lwt_log`](https://ocsigen.org/lwt/2.7.1/api/Lwt_log) might be deprecated in the future.
Therefore, we should move this repo to use the [`logs`](https://github.com/dbuenzli/logs) libray instead, which is the replacement recommended in the above thread. Its `Logs_lwt` library provides logging functions for `Lwt`.

Resources:
* [Manual for the `Lwt_log` module in Lwt 3.3.0](https://ocsigen.org/lwt/3.3.0/api/Lwt_log), the version we currently use
* [GitHub repo of the `logs` library](https://github.com/dbuenzli/logs)

### PR DESCRIPTION
**Pull request by psafont**
Fixes #13 and also prepares the code for Lwt 5.0

### CODE REVIEW

**Review by lindig — APPROVED** [review]
This looks good.

**Comment by psafont** [comment]
This needs changes in the daemonization of wsproxy (convert to systemd unit?)

**Comment by lindig** [comment]
I like that. I would assume that this is not difficult (looking at `wsproxy.ml`). But who spawns wsproxy currently and how would change that?

**Comment by lindig** [comment]
`wsproxy` is spawned from `xapi` in `console.ml` using Forkexecd.

**Comment by psafont** [comment]
My guess is that socket activation needs to be set up along with the service unit.

File: cli/wsproxy.ml [hunk 1]
```diff
@@ -45,25 +45,26 @@ let start path handler =
          >>= fun (fd_sock',_) ->
          (* Background thread per connection *)
          let _ : unit Lwt.t =
-           let buffer = String.make 16384 '\000' in
+           let buffer = Bytes.make 16384 '\000' in
            with_fd fd_sock'
              ~callback:(fun fd ->
-                let iov = Lwt_unix.io_vector ~buffer ~offset:0 ~length:16384 in
-                Lwt_unix.recv_msg ~socket:fd ~io_vectors:[iov])
+                 let io_vectors = Lwt_unix.IO_vectors.create () in
```
Comment (mseri) [inline cli/wsproxy.ml:43]: 🎊

File: cli/wsproxy.ml [hunk 2]
```diff
@@ -45,25 +45,26 @@ let start path handler =
...
+                 Lwt_unix.IO_vectors.append_bytes io_vectors buffer 0 16384;
```
Comment (rwdrich) [inline cli/wsproxy.ml:44]: Is it possible to keep the param names?

Comment (rwdrich) [reply cli/wsproxy.ml:44]: Apparently not :(

Comment (psafont) [reply cli/wsproxy.ml:44]: I tried [:(](https://github.com/ocsigen/lwt/blob/3e5923b40349e2b0b4c0820969dd81cad41b4719/src/unix/lwt_unix.cppo.ml#L692)

File: cli/wsproxy.ml [hunk 3]
```diff
@@ -13,7 +13,7 @@
  *)
 open Wslib
 
-let get_dir_path () = Printf.sprintf "/var/xapi/" 
+let get_dir_path () = Printf.sprintf "/var/xapi/"
```
Comment (gaborigloi) [inline cli/wsproxy.ml]: This could be changed into a variable? :sweat_smile:

**Review by gaborigloi — APPROVED** [review]

File: cli/wsproxy.ml (hunk 3 above)
Comment (psafont) [reply cli/wsproxy.ml]: I'm not even sure it's worth using a variable as it is, seeing that it's not configurable and it's only used in a single line.

**Review by gaborigloi — APPROVED** [review]

File: cli/wsproxy.ml [hunk 4]
```diff
@@ -117,18 +115,18 @@ let handler sock msg =
     let localhost = Unix.inet_addr_loopback in
     let addr = Unix.ADDR_INET(localhost, int_of_string sport) in
     proxy sock addr protocol
-  | _ -> Lwt_log.warning "Malformed msg: not proxying"
+  | _ -> Logs_lwt.warn (fun m -> m "Malformed msg: not proxying")
 
 
-let _ = 
+let _ =
+  Logs.set_reporter (Logs_fmt.reporter ());
```
Comment (gaborigloi) [inline cli/wsproxy.ml]: Do we need a different reporter for Lwt?
As suggested by 
https://erratique.ch/software/logs/doc/Logs_lwt/index.html#report_ex and
https://erratique.ch/software/logs/doc/Logs/index.html#setupreporter

Comment (psafont) [reply cli/wsproxy.ml]: We're debugging to `/dev/null` I don't think it matters if it's synchronous or not in this case.

**Comment by lindig** [comment]
What is the state of this? Can we merge this?

**Comment by psafont** [comment]
The socket/service are not activated when building the packages although they seem to work fine when enabled manually. It needs more changes in the spec files

**Comment by psafont** [comment]
Testing didn't go as planned: The connection to the guests are not successful when the wsproxy service is up and running, while managed by systemd.

I'll debug the changes in wsproxy as xapi is correctly opening the socket (wsproxy gets run by systemd)

**Comment by psafont** [comment]
Comparison of the log format before and after:

```
Oct 23 12:57:59 localhost wsproxy[26499]: main: About to start connection
Oct 23 12:57:59 localhost wsproxy[26499]: main: Got msg: hybi10:rfb:/var/run/xen/vnc-4
Oct 23 12:57:59 localhost wsproxy[26499]: main: New-style (hybi10) protocol
```
```
Oct 23 13:07:27 localhost wsproxy[17923]: wsproxy: [DEBUG] About to start connection
Oct 23 13:07:27 localhost wsproxy[17923]: wsproxy: [DEBUG] Got msg: hybi10:rfb:/var/run/xen/vnc-8
Oct 23 13:07:27 localhost wsproxy[17923]: wsproxy: [WARNING] Malformed msg: not proxying
```

Processing the message as malformed is not intended, have to look into it. The xapi plumbing looks good so far, however.

**Comment by psafont** [comment]
In order to use socket activation, we cannot create the sockets like we were doing before, we would have to use something like https://github.com/juergenhoetzel/ocaml-systemd

File: cli/wsproxy.ml [hunk 5]
```diff
@@ -13,17 +13,14 @@
...
 let with_fd = Lwt_support.with_fd
 
-let start path handler =
-  let dir_path = get_dir_path () in
-  let fd_sock_path = Printf.sprintf "%s%s" dir_path path in
-  Lwt_log.info_f "Starting wsproxy on %s" fd_sock_path >>= fun () ->
+let start handler =
+  let fd_sock_path = "/var/xapi/wsproxy" in
+  Logs_lwt.info (fun m -> m "Starting wsproxy on %s" fd_sock_path) >>= fun () ->
   let fd_sock = Lwt_unix.socket Unix.PF_UNIX Unix.SOCK_STREAM 0 in
```
Comment (edwintorok) [inline cli/wsproxy.ml]: Can we use Lwt_unix.stdin here?

Comment (psafont) [reply cli/wsproxy.ml]: yes, haven't had time to push the change ;p

**Review by rwdrich — APPROVED** [review]

**Pull request merged**